### PR TITLE
EE-8289: Simplify the logic in MultipleLastNames Generator

### PR DESCRIPTION
**Pull request by Tim-Hammonds**

### CODE REVIEW

File: src/test/java/uk/gov/digital/ho/pttg/application/namematching/candidates/MultipleLastNamesFunctionsTest.java [hunk 1]
```diff
@@ -20,11 +19,10 @@
 
     @Test
     public void addAllLastNameCombinationsMultipleNames() {
-        List<CandidateName> candidateNames = new ArrayList<>();
         List<String> firstNames = Arrays.asList("Aaa", "Bbb", "Ccc");
         List<String> lastNameCombinations = Arrays.asList("Ddd", "Eee", "Ddd Eee");
 
-        List<CandidateName> newCandidateNames = addAllLastNameCombinations(candidateNames, firstNames, lastNameCombinations);
+        List<CandidateName> newCandidateNames = addAllLastNameCombinations(firstNames, lastNameCombinations);
```
Comment (owainrobertsbjss) [inline src/test/java/uk/gov/digital/ho/pttg/application/namematching/candidates/MultipleLastNamesFunctionsTest.java]: As above - now that this method doesn't alter a list passed in as a parameter its name should change.

Comment (owainrobertsbjss) [reply src/test/java/uk/gov/digital/ho/pttg/application/namematching/candidates/MultipleLastNamesFunctionsTest.java]: Actually now I've read the test, it wasn't altering the list but the name is still poor.

File: src/test/resources/specs/name-matching/Full_stops_in_names.feature [hunk 2]
```diff
@@ -70,7 +70,7 @@ Feature: Names with full stops
     Then the following identities will be tried in this order
       | First name  | Last name | Date of Birth | nino      |
       | Aaa Bb. Ccc | Ddd       | 1987-12-10    | SE123456B |
-      | Bb. Ccc     | Ddd       | 1987-12-10    | SE123456B |
+      | Bb.         | Ddd       | 1987-12-10    | SE123456B |
```
Comment (owainrobertsbjss) [inline src/test/resources/specs/name-matching/Full_stops_in_names.feature]: I worry that this is a behaviour change - I guess the order doesn't matter but we must make sure we still have a test to prove that we use "St. John" etc. as a single word name in name matching.

Perhaps the issue is that Bb. is three characters. Changing the test so that the candidate's first name is B. Ccc might be the solution?

File: src/test/resources/specs/name-matching/Last_names_combination.feature [hunk 3]
```diff
@@ -120,7 +120,7 @@ Feature: Last name combinations
       | nino          | NR123456C     |
     Then the following names will be tried
       | First name    | Last name |
-      | Arthur Brian  | O Coates  |
+      | Arthur        | O Coates  |
```
Comment (owainrobertsbjss) [inline src/test/resources/specs/name-matching/Last_names_combination.feature]: It's only going to affect the edge case where the very first name is shorter than 3 characters but this change is maybe not one we should be making.

The Special Characters generator owns a Multiple Last Names strategy which used to add the entire name. I think perhaps it should now own a EntireNonAliasName generator so that we still get "Arthur Brian O Coates"?

File: src/test/resources/specs/name-matching/Full_stops_in_names.feature [hunk 4]
```diff
@@ -73,6 +73,20 @@ Feature: Names with full stops
       | Bb.         | Ddd       | 1987-12-10    | SE123456B |
     And a Matched response will be returned from the service
 
+  @name_matching
+  Scenario: Name when middle name has a full stop and a space and is matched with HMRC 2
```
Comment (owainrobertsbjss) [inline src/test/resources/specs/name-matching/Full_stops_in_names.feature]: I think the old test is now redundant / superseded by this one. Might have been your intention when you copied it here - sorry for not spotting when I was watching.

File: src/test/resources/specs/name-matching/Name_With_Special_Characters.feature [hunk 5]
```diff
@@ -73,13 +73,13 @@ Feature: Accept Hyphens and Apostrophes as part of the name matching
...
       | Bb-Ccc     | Aaa       | 1987-12-10    | SE 123456 B |
+      | Aaa        | BbCcc     | 1987-12-10    | SE 123456 B |
       | Aaa        | Bb Ccc    | 1987-12-10    | SE 123456 B |
     And a Matched response will be returned from the service
-    And HMRC was called 3 times
 
 
   @name_matching
-  Scenario: Applicant enters hyphen when name does not contain a hyphen
+  Scenario: Applicant enters hyphen when name does not contain a hyphen 2
```
Comment (owainrobertsbjss) [inline src/test/resources/specs/name-matching/Name_With_Special_Characters.feature]: Good spot that there were two tests with the same name!
